rfcs/80-curve-trait.md

```diff
@@ -0,0 +1,793 @@
+or distance.
+
+For example, we might use the `Curve` APIs to construct a `Curve<Transform>` in order to describe the motion 
+of an Entity with time:
+```rust
+// This is a `Curve<Quat>` that describes a rotation with time:
+let rotation_over_time = function_curve(interval(0.0, std::f32::consts::TAU).unwrap(), |t| Quat::from_rotation_z(t));
+// Here is a `Curve<Vec3>` that describes a translation with time:
+let translation_over_time = function_curve(interval(0.0, 1.0).unwrap(), |t| Vec3::splat(t));
+// Let's reparametrize the rotation so that it goes from 0 to 1 instead:
+let new_rotation = rotation_over_time.reparametrize_linear(interval(0.0, 1.0).unwrap()).unwrap();
+// Combine the two into a `Curve<(Vec3, Quat)>` by zipping them together:
+let translation_and_rotation = translation_over_time.zip(new_rotation).unwrap();
+// Join the two families of data to get a `Curve<Transform>`:
+let transform_curve = translation_and_rotation.map(|(t, r)| Transform::from_translation(t).with_rotation(r));
+```
+
+This could be used to actually set an entity's `Transform` in a system or otherwise:
+```rust
+*my_entity_transform = transform_curve.sample(0.6);
+```
+
+However, `Curve` is quite general, and it can be used for much more, including describing and manipulating
+easings, animations, geometry, camera moves, and more!
+
+The trait itself looks like this:
+
+```rust
+pub trait Curve<T>
+{
+    fn domain(&self) -> Interval;
+    fn sample(&self, t: f32) -> T;
+}
+```
+At a basic level, it encompasses values of type `T` parametrized over some particular `Interval` (its `domain`,
+allowing that data to be sampled by providing the associated time parameter. 
+
+### Intervals
+
+The `Interval` type plays a part of the definition of each `Curve`. This is a type which
+represents a nonempty closed interval over the `f32` numbers whose endpoints may be at infinity. Its provided methods are
+mostly self-explanatory:
+```rust
+/// Create a new [`Interval`] with the specified `start` and `end`. The interval can be infinite
+/// but cannot be empty; invalid parameters will result in an error.
+pub fn new(start: f32, end: f32) -> Result<Self, InvalidIntervalError> { //... }
+
+/// Get the start of this interval.
+pub fn start(self) -> f32 { //... }
+
+/// Get the end of this interval.
+pub fn end(self) -> f32 { //... }
+
+/// Get the length of this interval. Note that the result may be infinite (`f32::INFINITY`).
+pub fn length(self) -> f32 { //... }
+
+/// Returns `true` if this interval is finite.
+pub fn is_finite(self) -> bool { //... }
+
+/// Returns `true` if `item` is contained in this interval.
+pub fn contains(self, item: f32) -> bool { //... }
+
+/// Create an [`Interval`] by intersecting this interval with another. Returns an error if the
+/// intersection would be empty (hence an invalid interval).
+pub fn intersect(self, other: Interval) -> Result<Interval, InvalidIntervalError> { //... }
```

Comment:
Follow-up: I think we want a union method too, but that will need a new error arm for non-contiguous intervals.

Reply:
Yeah, that sounds useful.